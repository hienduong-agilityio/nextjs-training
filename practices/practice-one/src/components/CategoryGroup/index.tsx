'use client';

// Libraries
import { usePathname, useSearchParams } from 'next/navigation';
import { twMerge } from 'tailwind-merge';

// Components
import Link from 'next/link';

// Helpers
import { capitalizeText } from '@/helpers';

// Constants
import { ALL_CATEGORIES, SEARCH_PARAMS } from '@/constants';

interface ICategoryGroupProps {
  title: string;
  items: { name: string; count: number }[];
}

export function CategoryGroup({ title, items }: ICategoryGroupProps) {
  const pathname = usePathname();
  const searchParams = useSearchParams();

  const currentCategory =
    searchParams.get(SEARCH_PARAMS.CATEGORY) ?? ALL_CATEGORIES.ALL;

  const handleCategoryClick = (itemName: string) => {
    const updatedParams = new URLSearchParams();

    // Update the category query parameter
    if (itemName !== ALL_CATEGORIES.ALL) {
      updatedParams.set(SEARCH_PARAMS.CATEGORY, itemName);
    }

    return `${pathname}?${updatedParams.toString()}`;
  };

  return (
    <div className="mb-6 bg-secondary-100 p-4 rounded">
      <h3 className="text-lg font-bold mb-4">{title}</h3>
      <ul className="space-y-2 text-sm text-gray-700">
        {items.map((item) => {
          const isActive =
            item.name === ALL_CATEGORIES.ALL
              ? currentCategory === ALL_CATEGORIES.ALL || !currentCategory
              : item.name === currentCategory;

          return (
            <li className="flex justify-between" key={item.name}>
              <Link
                href={handleCategoryClick(item.name)}
                className={twMerge(
                  'w-full block font-semibold py-2 rounded hover:bg-secondary-200',
                  isActive
                    ? 'bg-primary-500 text-primary-200'
                    : 'text-gray-800',
                )}
              >
                {capitalizeText(item.name)}
              </Link>
            </li>
          );
        })}
      </ul>
    </div>
  );
}
