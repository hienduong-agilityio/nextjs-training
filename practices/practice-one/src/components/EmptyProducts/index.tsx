import Link from 'next/link';

export const EmptyProducts = ({
  message,
  actionLabel,
  actionHref,
}: {
  message: string;
  actionLabel: string;
  actionHref: string;
}) => (
  <div className="flex flex-col items-center gap-4 py-16">
    <p className="text-xl text-gray-600">{message}</p>
    <Link
      href={actionHref}
      className="inline-block px-6 py-2 text-white bg-blue-500 rounded-lg"
    >
      {actionLabel}
    </Link>
  </div>
);
