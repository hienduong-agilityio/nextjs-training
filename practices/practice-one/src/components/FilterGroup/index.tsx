interface FilterGroupProps {
  title: string;
  items: { name: string; count: number }[];
}

export const FilterGroup = ({ title, items }: FilterGroupProps) => {
  return (
    <div className="mb-6 bg-secondary-100 p-4 rounded">
      <h3 className="text-lg font-bold mb-4">{title}</h3>
      <ul className="space-y-2 text-sm text-gray-700">
        {items.map((item) => (
          <li key={item.name} className="flex justify-between py-2">
            <button className="font-semibold hover:text-primary-200 ">
              {item.name}
            </button>
            <span>({item.count})</span>
          </li>
        ))}
      </ul>
    </div>
  );
};
