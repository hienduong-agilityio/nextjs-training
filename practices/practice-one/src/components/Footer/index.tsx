// Icon
import {
  FacebookIcon,
  LogoIcon,
  MasterCardIcon,
  PaypalIcon,
  TwitterIcon,
  VisaIcon,
  WesternIcon,
} from '@/icons';

// Mock
import { FOOTER_NAVIGATION } from '@/mocks';

// Components
import Link from 'next/link';

// Constants
import { ROUTE } from '@/constants';

export const Footer = () => {
  return (
    <footer className="w-full bg-primary-50">
      <div className="md:container px-4 pt-6 md:mx-auto sm:px-6 md:px-8 lg:px-12 xl:px-16 sm:pt-8 md:pt-12 lg:pt-16 xl:pt-20">
        <div className="py-8">
          {/* Main Footer Content */}
          <div className="flex flex-col gap-10 lg:flex-row lg:justify-between">
            {/* Logo Section */}
            <div className="max-w-[290px] flex flex-col gap-4">
              <Link href={ROUTE.PRODUCTS} className="flex items-center gap-2">
                <LogoIcon aria-label="logo" />
                <span className="text-lg font-bold sm:text-xl lg:text-2xl">
                  E-Comm
                </span>
              </Link>
              <span className="text-sm font-normal leading-normal">
                Lorem Ipsum is simply dummy text of the printing and typesetting
                industry. Lorem Ipsum has been the industry standard dummy text
                ever. Since the 1500s, when an unknown printer.
              </span>
            </div>

            <div className="flex flex-col w-full gap-8 sm:flex-row sm:justify-evenly lg:w-3/4">
              {/* Follow Us */}
              <div className="lg:max-w-[220px] flex flex-col gap-4">
                <h4 className="font-medium text-foreground-400">Follow Us</h4>
                <span className="text-sm">
                  Since the 1500s, when an unknown printer took a galley of type
                  and scrambled.
                </span>
                <div className="flex gap-4">
                  <Link href="https://www.facebook.com/" aria-label="Facebook">
                    <FacebookIcon color="#385C8E" />
                  </Link>
                  <Link href="https://www.twitter.com/" aria-label="Twitter">
                    <TwitterIcon />
                  </Link>
                </div>
              </div>

              {/* Contact Us */}
              <div className="lg:max-w-[150px] flex flex-col gap-4">
                <h4 className="font-medium text-foreground-400">Contact Us</h4>
                <address className="text-sm not-italic">
                  E-Comm, 4578 Marmora Road, Glasgow D04 89GR
                </address>
              </div>
            </div>
          </div>

          {/* Footer Links */}
          <div className="grid grid-cols-1 gap-8 pt-16 sm:grid-cols-2 lg:grid-cols-4">
            {FOOTER_NAVIGATION.map((item) => (
              <ul key={item.title}>
                <li>
                  <h4 className="mb-5 font-medium text-foreground-400">
                    {item.title}
                  </h4>
                  <ul>
                    {item.items.map((item) => (
                      <li
                        key={item.name}
                        className="mb-3 text-sm text-content2 hover:text-primary-700"
                      >
                        <Link href={item.url} className="hover:underline">
                          {item.name}
                        </Link>
                      </li>
                    ))}
                  </ul>
                </li>
              </ul>
            ))}
          </div>
        </div>

        {/* Footer Bottom */}
        <div className="flex flex-col items-center justify-between gap-6 py-6 border-t-2 border-secondary-100 lg:flex-row ">
          <p className="text-sm text-secondary-500">
            &#169; 2018 Ecommerce theme by www.bisenbaev.com
          </p>
          <div className="flex gap-2">
            <WesternIcon aria-label="Western" />
            <MasterCardIcon aria-label="MasterCard" />
            <PaypalIcon aria-label="Paypal" />
            <VisaIcon aria-label="Visa" />
          </div>
        </div>
      </div>
    </footer>
  );
};
