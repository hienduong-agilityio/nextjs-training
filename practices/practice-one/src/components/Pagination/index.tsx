'use client';

// Libraries
import { twMerge } from 'tailwind-merge';

// Hooks
import { usePathname, useSearchParams } from 'next/navigation';

// Components
import Link from 'next/link';

interface PaginationProps {
  totalPages: number;
  currentPage?: number;
}

export function Pagination({ totalPages, currentPage }: PaginationProps) {
  const pathname = usePathname();
  const searchParams = useSearchParams();

  const page = (currentPage ?? Number(searchParams.get('page'))) || 1;

  const createPageURL = (page: number) => {
    const params = new URLSearchParams(searchParams);

    params.set('page', page.toString());

    return `${pathname}?${params.toString()}`;
  };

  // Generate all pages
  const pages = Array.from({ length: totalPages }, (_, i) => i + 1);

  return (
    <div className="flex items-center space-x-2" aria-label="Pagination">
      {pages.map((pageItem) => (
        <Link
          key={pageItem}
          href={createPageURL(pageItem)}
          className={twMerge(
            'px-5 py-3 ',
            page === pageItem
              ? 'bg-primary-200 text-white border-primary-200'
              : 'text-gray-700 hover:bg-gray-100',
          )}
          aria-current={page === pageItem ? 'page' : undefined}
        >
          {pageItem}
        </Link>
      ))}
    </div>
  );
}
