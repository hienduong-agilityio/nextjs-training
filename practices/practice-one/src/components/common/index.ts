export { default as Button } from './Button';
export { default as InputField } from './InputField';
export { default as Tabs } from './Tabs';
export { default as Select } from './Select';
export { default as IconButton } from './IconButton';
export * from './IconButton';
