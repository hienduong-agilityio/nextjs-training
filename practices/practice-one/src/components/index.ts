export * from './Header';
export { default as SearchBox } from './SearchBox';
export * from './SearchBox';
export * from './ProductTab';

// Common
export * from './common';
