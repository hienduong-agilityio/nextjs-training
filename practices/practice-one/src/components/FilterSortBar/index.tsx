'use client';

// Hooks
import { useSearchParams, useRouter } from 'next/navigation';
import { useEffect, useState } from 'react';

// Components
import { Select } from '@/components';

// Constants
import { SEARCH_PARAMS, SORT_PRODUCT_OPTIONS } from '@/constants';

interface FilterSortBarProps {
  itemCount: number;
  sortOptions: string[];
  showOptions: string[];
}

export const FilterSortBar = ({
  itemCount = 0,
  sortOptions = SORT_PRODUCT_OPTIONS,
  showOptions = ['6'],
}: FilterSortBarProps) => {
  const searchParams = useSearchParams();
  const router = useRouter();

  const [selectedLimit, setSelectedLimit] = useState<string>(showOptions[0]);
  const [selectedSort, setSelectedSort] = useState<string>(sortOptions[0]);

  // TODO: Remove useEffect.
  useEffect(() => {
    // Get query parameters from the URL
    const limitParam = searchParams.get(SEARCH_PARAMS.LIMIT) ?? showOptions[0];
    const sortParam = searchParams.get(SEARCH_PARAMS.SORT_BY) ?? sortOptions[0];

    setSelectedLimit(limitParam);
    setSelectedSort(sortParam);
  }, [searchParams, showOptions, sortOptions]);

  const updateQueryParams = (param: string, value: string) => {
    const currentParams = new URLSearchParams(searchParams.toString());
    currentParams.set(param, value);

    // Reset pagination if a filter changes
    if (param === SEARCH_PARAMS.LIMIT) {
      currentParams.delete(SEARCH_PARAMS.PAGE);
    }

    router.push(`?${currentParams.toString()}`);
  };

  const handleShowChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const value = event.target.value;

    setSelectedLimit(value);
    updateQueryParams(SEARCH_PARAMS.LIMIT, value);
  };

  const handleSortChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const value = event.target.value;

    setSelectedSort(value);
    updateQueryParams(SEARCH_PARAMS.SORT_BY, value);
  };

  return (
    <div className="bg-secondary-100 p-4 shadow-md md:sticky rounded top-0 z-30 w-full">
      <div className="flex flex-col sm:flex-row sm:justify-between sm:items-center space-y-4 sm:space-y-0">
        {/* Item count */}
        <div className="text-sm w-max text-gray-600">{itemCount} Items</div>

        {/* Filters */}
        <div className="flex flex-col sm:flex-row sm:items-center sm:space-x-4 space-y-4 sm:space-y-0">
          <div className="flex gap-2 items-center">
            <span>Sort</span>
            {/* TODO: Custom style for select arrow by tailwinds classes*/}
            <Select
              options={sortOptions}
              value={selectedSort}
              customClass="w-auto bg-secondary-100 capitalize"
              onChange={handleSortChange}
            />
          </div>
          <div className="flex gap-2 items-center">
            <span>Show</span>
            <Select
              options={showOptions}
              value={selectedLimit}
              customClass="w-auto bg-secondary-100"
              onChange={handleShowChange}
            />
          </div>
        </div>
      </div>
    </div>
  );
};
