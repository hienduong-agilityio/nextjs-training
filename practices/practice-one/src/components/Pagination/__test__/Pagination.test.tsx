// Libraries
import { render, screen, fireEvent } from '@testing-library/react';
import { usePathname, useSearchParams } from 'next/navigation';

// Components
import { Pagination } from '@/components';

// Constants
import { ROUTE } from '@/constants';

jest.mock('next/navigation', () => ({
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
}));

jest.mock('react', () => {
  const actualReact = jest.requireActual('react');
  return {
    ...actualReact,
    useOptimistic: jest.fn((initialValue) => [initialValue, jest.fn()]),
  };
});

describe('Pagination Component', () => {
  const mockUsePathname = usePathname as jest.Mock;
  const mockUseSearchParams = useSearchParams as jest.Mock;

  beforeEach(() => {
    mockUsePathname.mockReturnValue(ROUTE.COLLECTION);
    mockUseSearchParams.mockReturnValue(new URLSearchParams());
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderPagination = (
    props: Partial<React.ComponentProps<typeof Pagination>> = {},
  ) => render(<Pagination totalPages={5} {...props} />);

  const verifyPageAttributes = (
    page: HTMLElement,
    pageNumber: number,
    isActive: boolean,
  ) => {
    expect(page).toHaveTextContent(pageNumber.toString());
    expect(page).toHaveAttribute(
      'href',
      `${ROUTE.COLLECTION}?page=${pageNumber}`,
    );
    if (isActive) {
      expect(page).toHaveClass('bg-primary-200 text-white border-primary-200');
      expect(page).toHaveAttribute('aria-current', 'page');
    } else {
      expect(page).toHaveClass('text-gray-700 hover:bg-gray-100');
      expect(page).not.toHaveAttribute('aria-current');
    }
  };

  const getPageLinks = () => screen.getAllByRole('link');

  it('renders all pages correctly', () => {
    const { asFragment } = renderPagination();

    const pages = getPageLinks();
    expect(pages).toHaveLength(5);

    pages.forEach((page, index) => {
      verifyPageAttributes(page, index + 1, index === 0);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('applies active styles to the current page', () => {
    const { asFragment } = renderPagination({ currentPage: 3 });

    const pages = getPageLinks();

    const pageLinks = getPageLinks();
    const targetPage = pageLinks[3];
    fireEvent.click(targetPage);

    verifyPageAttributes(targetPage, 4, false);

    pages.forEach((page, index) => {
      verifyPageAttributes(page, index + 1, index + 1 === 3);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('constructs correct URLs for each page with query parameters', () => {
    mockUseSearchParams.mockReturnValue(
      new URLSearchParams({ search: 'query' }),
    );
    const { asFragment } = renderPagination();

    const pages = getPageLinks();
    pages.forEach((page, index) => {
      const expectedHref = `${ROUTE.COLLECTION}?search=query&page=${index + 1}`;
      expect(page).toHaveAttribute('href', expectedHref);
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
