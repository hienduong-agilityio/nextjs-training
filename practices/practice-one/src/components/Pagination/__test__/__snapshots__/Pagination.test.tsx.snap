// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component applies active styles to the current page 1`] = `
<DocumentFragment>
  <div
    aria-label="Pagination"
    class="flex items-center space-x-2"
  >
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?page=1"
    >
      1
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?page=2"
    >
      2
    </a>
    <a
      aria-current="page"
      class="px-5 py-3 bg-primary-200 text-white border-primary-200"
      href="/collection?page=3"
    >
      3
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?page=4"
    >
      4
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?page=5"
    >
      5
    </a>
  </div>
</DocumentFragment>
`;

exports[`Pagination Component constructs correct URLs for each page with query parameters 1`] = `
<DocumentFragment>
  <div
    aria-label="Pagination"
    class="flex items-center space-x-2"
  >
    <a
      aria-current="page"
      class="px-5 py-3 bg-primary-200 text-white border-primary-200"
      href="/collection?search=query&page=1"
    >
      1
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?search=query&page=2"
    >
      2
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?search=query&page=3"
    >
      3
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?search=query&page=4"
    >
      4
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?search=query&page=5"
    >
      5
    </a>
  </div>
</DocumentFragment>
`;

exports[`Pagination Component renders all pages correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Pagination"
    class="flex items-center space-x-2"
  >
    <a
      aria-current="page"
      class="px-5 py-3 bg-primary-200 text-white border-primary-200"
      href="/collection?page=1"
    >
      1
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?page=2"
    >
      2
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?page=3"
    >
      3
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?page=4"
    >
      4
    </a>
    <a
      class="px-5 py-3 text-gray-700 hover:bg-gray-100"
      href="/collection?page=5"
    >
      5
    </a>
  </div>
</DocumentFragment>
`;
